### ISSUE TITLE
Sandbox slowness on OSX

### ISSUE THREAD

**Issue by fenghaolw** [bazelbuild/bazel#8230]
### Description of the problem / feature request:

building has been extremely slow with the default darwin-sandbox

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

A mini repro could be found in https://github.com/alexeagle/rules_sass_repro
The repro contains 40 empty sass files
Running the sass compiler on them should be fast

bazel build :all takes ~60s on my mac

bazel build --strategy=SassCompiler=local :all takes ~4s

### What operating system are you running Bazel on?

Mac OS 10.14.4

### What's the output of `bazel info release`?

release 0.25.0

###  Have you found anything relevant by searching the web?

I found these issues: https://github.com/bazelbuild/bazel/issues/902 and https://github.com/bazelbuild/bazel/issues/1836 but they all seem obsolete.

**Comment by fenghaolw** [comment]
Note that on linux I also see slowness for the sandbox, but much better than mac.

It takes ~15s for linux-sandbox, and ~3s for strategy=SassCompiler=local on my Lenovo P920.

**Comment by meisterT** [comment]
cc @jmmv

**Comment by alexeagle** [comment]
discussed with @dslomov this morning, I think we should disable sandbox on mac by default. It has dubious benefits for keeping your build hermetic while causing lots of broken (non-hermetic) and slow builds

**Comment by dslomov** [comment]
ping @jmmv @philwo for opinions.

**Comment by philwo** [comment]
> I think we should disable sandbox on mac by default

That's a bold strategy. Because without sandboxing, all actions will run in the same, shared `execroot`, which contains all files from your workspace and all output files, missing dependency declarations will no longer be detected. This can cause Bazel to assume that targets that should be rebuild are still up to date, which can result in wrong build results.

If you think that this doesn't matter for your particular use case or if you have other mitigations for this, then feel free to go ahead with the plan, but in general this is not recommended.

If your sass build steps are *that* slow, it can only mean that they have a huge number of input files. Is that the case?

**Comment by Globegitter** [comment]
I have also noticed that sandboxing is generally slower on macOS compared to linux. Does anyone actually have any explanation of why that is? Is it due to some macOS reasons, or is the sandbox on macOS not as optimized?

Also @fenghaolw have you tested the same build with https://github.com/bazelbuild/sandboxfs? Apparently there is a chance it can make sandboxed builds on macOS quite a bit faster. Would be curios to see some comparisons with that.

**Comment by keith** [comment]
We tested sandboxfs with our iOS build without much luck https://github.com/bazelbuild/sandboxfs/issues/76

**Comment by jmmv** [comment]
There are two problems regarding sandboxing slowness:

1. We have to create possibly thousands of symlinks per action, and that turns out to degrade badly on macOS. Microbenchmarks don't seem to expose this behavior easily though. sandboxfs is supposed to get rid of this problem but hasn't been successful for all builds yet (there is room for optimizations there that I haven't had a chance to tackle).

2. There is the theory that `sandbox-exec` is slow. Same as above, microbenchmarks don't seem to confirm this -- but I haven't looked into it with great detail. If this is inherently slow, then there is nothing much we can do (especially considering that this is deprecated).

There is this other idea floating around of changing the way sandboxing works: instead of preventing operations, we just log them all and then verify, after the action completes, that whatever the action accessed was allowed. This would avoid the symlinks completely and also bypass sandbox-exec. I think this is how Windows sandboxing will work. But... I don't think there is a way to do this kind of tracing on macOS out of the box that doesn't require either root or disabling SIP. It's worth investigating though.

**Comment by Globegitter** [comment]
@keith Yeah I saw that, very interesting. I have just done some quick tests on my 6 year old macbook Pro and for a tiny JS build at least (with already a fair amount of inputs due to npm modules) my build time halved from roughly 6s to 3s. Without sandbox it is 2s. That does look pretty good to me but I do wonder for what type of builds and what type of setup/configuration that differs.

**Comment by jmmv** [comment]
Some tests I ran with this reproduction:

```
$ bazel build :all
INFO: Elapsed time: 36.556s, Critical Path: 7.71s
$ bazel build --spawn_strategy=local :all
INFO: Elapsed time: 5.569s, Critical Path: 1.23s
$ bazel build --experimental_use_sandboxfs :all
INFO: Elapsed time: 9.479s, Critical Path: 3.16s
$ bazel build --sandbox_debug :all
INFO: Elapsed time: 18.091s, Critical Path: 3.90s
$ bazel build --experimental_sandbox_async_tree_delete_idle_threads=auto :all
INFO: Elapsed time: 23.156s, Critical Path: 4.75s
```

And corresponding observations:

1. Each action in this build has 11k files.
1. sandboxfs does seem to help (as expected based on the previous).
1. `--sandbox_debug` makes quite a bit of a difference. Deleting all the symlink trees is expensive, and this flag has the side-effect of not deleting them. But creating them is also quite expensive.
1. The new `--experimental_sandbox_async_tree_delete_idle_threads=auto` helps approximate the behavior of `--sandbox_debug` and seems like a significant improvement over the current behavior. We should enable this new feature by default, but I remember seeing a crash recently that needs investigation...

**Comment by Globegitter** [comment]
@jmmv I have also seen improvements with the async delete and I reported the crash here: https://github.com/bazelbuild/bazel/issues/7527#issuecomment-491634909

**Comment by jmmv** [comment]
Thanks. I have a fix for that crash when enabling asynchronous deletions but I'm still having trouble with one test... plus in benchmarking the change, I'm consistently getting worse results in large builds. Need to look into that too.

Separately, I've been looking into the old claim that `sandbox-exec` is inherently slow. I don't think it is. Some microbenchmarks showed no slowdowns a long time ago. I have now patched the "local" strategy to run everything through `sandbox-exec` using a configuration file that matches what the "sandboxed" strategy generates and got the following for a large iOS build:

```
local-head: mean 2074.50, median 2075.00, stddev 6.69
local-new: mean 2109.00, median 2110.50, stddev 5.15
```

where `local-head` is the behavior of the "local" strategy with head Bazel and `local-new` is the same strategy patched as described above. And yes, there seems to be a little bit of overhead, but the differences between the two are minimal.

Therefore I conclude that the major cost of sandboxing in macOS continues to be the symlinks tree and handling files.

**Comment by meteorcloudy** [comment]
@alexeagle Since we already have a P1 bug for Mac Sandbox performance issue, I think we can continue our previous discussion here.

The current problem is sandbox on Mac is very slow, especially for Angular project, where there are  a large number of input files. So we wonder that's the way forward.

@jmmv Is there any current work on improving the Mac sandbox performance? From the above discussion, sandbox might be a way, how big could the improvement be? @alexeagle Did you every try it before?

If it's hard to improve the performance, should we disable sandbox on Mac? Disabling sandbox could make the build less hermetic, but @alexeagle also mentioned it was causing some broken build that may not be the fault of the build rules? I'm also wondering how this would affect remote execution. Some issues may only be spotted when remote execution is enabled if we didn't have sandbox in advance.

If we should not disable sandbox on mac, we should provide a nicer way to disable sandbox for a specific platform. That means, we only want to apply flags like `--spawn_strategy=standalone` on Mac. Currently, you can group those flags under a config in bazelrc file. But you still have to enable the config from a bazel command line invocation. To make things easier, we can have a platform-default config, like:
```
build:windows --spawn_strategy=standalone
build:linux  --spawn_strategy=sandboxed
build:macos  --spawn_strategy=standalone
```
Then Bazel loads different flags according to the platform it's running on and users never have to pass `--config=macos` explicitly.

**Comment by Globegitter** [comment]
Irregardless of the performance improvements of the sandbox I have had the need before for platform specific flags and it would indeed be very nice if it could just be defined centrally and the user would not have to pass in a `- -config` flag.

**Comment by meteorcloudy** [comment]
I agree, https://github.com/bazelbuild/bazel/issues/5055 is for tracking this feature request. It has been silent for a while, I'll follow up on this.

**Comment by DavidGoldman** [comment]
For what it's worth, Swift users will have issues debugging with sandboxing enabled (see https://github.com/bazelbuild/tulsi/issues/15)

**Comment by acecilia** [comment]
Could the slowness be due to notarization? https://sigpipe.macromates.com/2020/macos-catalina-slow-by-design/

Seems like since notarization was introduced, all scripts that run for the first time are firing a network request to apple. Since bazel uses sandbox, potentially all scripts there are considered new by the system.

I could not check this theory though

**Comment by meisterT** [comment]
While this is a pain point for users, there doesn't seem to be low hanging fruits and we don't have the bandwidth at the moment to work on this. Contributions welcome!

**Comment by larsrc-google** [comment]
@acecilia That might be the case for genrules, maybe, but it wouldn't affect Javac and other pre-built binaries more than at most once per build. It would affect local execution even worse, though, so I don't think that's to blame. Also, I'm not seeing that notarization slowdown on my (Catalina) machine, regardless of security setting.

**Comment by jakeleventhal** [comment]
Bump on this

**Comment by alexeagle** [comment]
I think this problem is especially noticeable under nodejs rules, where the number of inputs is easily an order of magnitude more than most other ecosystems due to lack of archive files (every file in the package is a separate input) and the dependency problem in JS (hundreds of transitive dependencies for common tools like react-scripts).
We're working on a fix in rules_nodejs to provide each package as a directory (TreeArtifact) instead, though it's a breaking change (you can no longer reference individual files with labels)

**Comment by philwo** [comment]
I have recently tried replacing symlinks with APFS copy-on-write copies and also putting Bazel's output_base on a separate APFS volume, unfortunately it wasn't a clear win and sometimes even slower (although it should be much better for hermeticity).

Next thing I wanted to try was whether it's faster to keep persistent sandboxing directories around and only apply the changes required (e.g. remove files that the new action no longer needs, add missing files that it needs and ensure that all existing files haven't been modified). A bit like how "rsync -av --delete" works.

Reducing the number of files is a good way to improve performance, so @alexeagle's plan sgtm for this specific case.

**Comment by larsrc-google** [comment]
At
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/worker/WorkerExecRoot.java#L148,
the worker sandbox deletes files that shouldn't be in the sandbox for the
action, but it doesn't remove the inputs that are already there from
`inputsToCreate`. I can't tell if there's a reason for that, nor whether
re-creating them takes significant time.

-Lars

**Comment by philwo** [comment]
> I can't tell if there's a reason for that, nor whether re-creating them takes significant time.

The main reason for that special behavior for the sandboxed persistent worker case was that as the worker process stays running, we cannot delete files that it already has opened on Windows, so just removing and recreating everything doesn't work.

However, I think only making the required changes is much faster than deleting and recreating everything, too. For example, on macOS the Spotlight indexer tries to index all the sandboxes of Bazel all the time and also fseventsd keeps track of all the ongoing modifications and needs a considerable amount of CPU time for that.

**Comment by larsrc-google** [comment]
I wasn't suggesting recreating everything, rather the opposite: We should
take existing files out of inputsToCreate. I just don't know if there's
some reason that's not already done.

-Lars

**Comment by philwo** [comment]
Ohhh! Sorry, now I understand. Maybe we did that because we couldn't (or just didn't?) verify that the input files were unmodified? We definitely want to avoid the case where a sandboxed process modifies the input files and then those get reused by a following action.

**Comment by larsrc-google** [comment]
@fenghaolw Could you try running with the `--experimental_reuse_sandbox_directories` flag and see if that speeds up the sandboxing sufficiently?

**Comment by fenghaolw** [comment]
I no longer work on bazel rules. @alexeagle would you mind check if that flag helps bazel rules performance?

**Comment by alexeagle** [comment]
Ooh hi @larsrc-google that flag sounds magical if it does what I imagine it does! For nodejs projects this flag might obviate the need for a big redesign we've been doing for exposing input artifacts as directories.
 I'm on vacation for a couple weeks but will try to repro this after.

**Comment by lencioni** [comment]
We have been working to speed up sandboxing for our TypeScript Bazel build, which was originally often timing out at a 90 minute limit and frequently running at ~60 minutes (currently running without caching or remote build execution). We had identified the primary culprit as sandboxing slowness, which we observed both on Mac OS (our laptops) and Linux (our CI machines).

We had previously only enabled the new rules_nodejs `exports_directories_only` in our `yarn_install`, which dropped our TypeScript build down to 33-38 minutes, with occasional spikes of 55 minutes.

![image](https://user-images.githubusercontent.com/195534/127496566-162d89bc-0890-4e36-a9cb-23b8434ab238.png)

Yesterday, I tried adding both `--experimental_reuse_sandbox_directories` and `--experimental_sandbox_async_tree_delete_idle_threads=1` in our build and it seems to have a good impact on top of the rules_nodejs `exports_directories_only` option. For our TypeScript build, this brings things down from 33-38 minutes with spikes of ~55 minutes to about 22-24 minutes steady (so far).

![image](https://user-images.githubusercontent.com/195534/127495717-34800826-84d8-4353-ba9e-90a0de661250.png)

Zooming out for context, here are the max and average runtime trends for these jobs over the past 13 weeks, which includes all of these changes:

![image](https://user-images.githubusercontent.com/195534/127497011-4cf4fc08-9b50-4abc-b944-e29323fd524f.png)

**Comment by alexeagle** [comment]
@lencioni I'm so glad you found time to try this out, as I haven't yet.

Did you do any experiments yet with `experimental_reuse_sandbox_directories` but not enabling `exports_directories_only` ?

**Comment by lencioni** [comment]
@alexeagle I ran out of time yesterday to do that, but I'll kick off some test builds now and report back when I can.

**Comment by purkhusid** [comment]
For those using LTS releases, will these flags be available in 4.2 or only in 5.0?

**Comment by meisterT** [comment]
@larsrc-google is working on worker related cherrypicks for 4.2 - not sure if that CL will be included

**Comment by larsrc-google** [comment]
I was not planning to cherry-pick `experimental_reuse_sandbox_directories` for 4.2, no. It's new and experimental. The `experimental_sandbox_async_tree_delete_idle_threads` option is years old.

**Comment by lencioni** [comment]
I've run the build 3 times now, with `experimental_reuse_sandbox_directories` and `experimental_sandbox_async_tree_delete_idle_threads` enabled, but `exports_directories_only` disabled. These builds clock in at 28m 14s, 30m 15s, and 28m 29s. So about 6 minutes slower.

**Comment by jakeleventhal** [comment]
Any updates on this? My team is waiting on this before moving to bazel

**Comment by keith** [comment]
I don't think there's really a way out of this in the short term. I would suggest disabling sandboxing for local dev for performance, and enabling it for CI release builds in case anything slips through. This is what most folks are doing today.

**Comment by jakeleventhal** [comment]
@keith this sucks but i guess it works. what is the long term eta on this?

**Comment by philwo** [comment]
@jakeleventhal We would love to make it faster, but I think this would require a fundamental redesign of the implementation on macOS. So far we've struggled to find a better mechanism in macOS than using `sandbox-exec`, with which we could implement faster and reliable sandboxing. The system just doesn't seem to provide any good APIs for this. If you know of any or tools which implement sandboxing on macOS, please send us pointers!

The most advanced sandboxing engine I know of is part of Microsoft's BuildXL: https://github.com/microsoft/BuildXL/blob/master/Documentation/Specs/Sandboxing.md#macos-sandboxing but considering its complexity, so far no one has dared to look into if / how we could use it for Bazel.

**Comment by ob** [comment]
FWIW, I don't think `sandbox-exec` is the bottleneck. It's just that the filesystem is much slower in creating/deleting files and symlinks than the linux filesystem. I tried the original test case submitted (clone https://github.com/alexeagle/rules_sass_repro)  with bazel as of e4a26b6391, and this is what I got:

```
$ bazel build --incompatible_restrict_string_escapes=false :all
...
INFO: Analyzed 40 targets (140 packages loaded, 4664 targets configured).
INFO: Found 40 targets...
INFO: Elapsed time: 58.295s, Critical Path: 21.33s
INFO: 46 processes: 6 internal, 40 darwin-sandbox.
INFO: Build completed successfully, 46 total actions```

vs 

```
$ bazel build --incompatible_restrict_string_escapes=false --strategy=SassCompiler=local :all
...
INFO: Analyzed 40 targets (140 packages loaded, 4664 targets configured).
INFO: Found 40 targets...
INFO: Elapsed time: 14.365s, Critical Path: 2.20s
INFO: 46 processes: 6 internal, 40 local.
INFO: Build completed successfully, 46 total actions
```

If I use `--experimental_reuse_sandbox_directories`, I get:

```
$ bazel build --incompatible_restrict_string_escapes=false --experimental_reuse_sandbox_directories :all
...
INFO: Analyzed 40 targets (140 packages loaded, 4664 targets configured).
INFO: Found 40 targets...
INFO: Elapsed time: 20.694s, Critical Path: 9.19s
INFO: 46 processes: 6 internal, 40 darwin-sandbox.
INFO: Build completed successfully, 46 total actions```
```

So it definitively helps.

**Comment by ob** [comment]
I also got a trace profile:

<img width="1255" alt="image" src="https://user-images.githubusercontent.com/4950/133649996-474cd412-d288-447d-a1a5-6133b2bdac5f.png">

as you can see the sandbox setup/teardown is where most of the time is spent. This is without `--experimental_reuse_sandbox_directories`

**Comment by jmmv** [comment]
Agree with what @ob said. When I was looking into this, I patched Bazel to run all actions via `sandbox-exec` but *without* doing any of the file system work--and the performance overhead on various builds I tried was negligible.

**Comment by philwo** [comment]
Yes, you are correct. Our current sandbox  mechanism needs these containers on disk. We have experimented with various replacements and optimizations (sandboxfs, APFS clonefile copies), but they either don’t help or have other caveats.

More advanced sandboxing mechanisms like BuildXL‘s sandbox would not require building file system based sandbox containers. Instead they observe and filter the syscalls and file accesses via kernel mechanisms and thus avoid this overhead. However, this filtering of course comes with a different kind of overhead which might be even worse, depending on the workload.

Also note that sandbox-exec is deprecated: https://stackoverflow.com/questions/56703697/how-to-sandbox-third-party-applications-when-sandbox-exec-is-deprecated-now - but due to the widespread usage I hope they aren’t really considering to remove it.

The situation is not great. :/ Any help (even just ideas) to improve it is of course welcome!

**Comment by burdiyan** [comment]
In my experience the major cause of Bazel being slow on macOS is that symlinks on macOS are generally slow.

For example I have a genrule target with large number of source files with nested directories (basically the whole project, written in Go) and I do plain `go build` inside the execroot, and it takes about 50 seconds to build. When I do the same locally without Bazel it takes 5 seconds. And, the important thing is that if inside the same genrule target I just `cd` out of the execroot into the actual workspace tree, and do the same `go build` it takes 6-7 seconds.

So, my intuition says that the overhead is not in Bazel trying to create the symlink forest and prepare the sandbox, but in the fact that reading symlinked files is just slow on macOS. Because in both cases Bazel had to build the symlinks anyways.

Also, FWIW, the [Please](https://please.build) build-system which is inspired by Bazel just copies input files into the execroot instead of symlinking. I still have to try how much of an improvement that brings, but I think it's quite a lot

**Comment by burdiyan** [comment]
Indeed, I have replicated the same target in Please, and it takes 9 seconds in Please what takes 58 seconds in Bazel. Both systems give you the same benefits in terms of input files tracking, but Bazel uses symlinks (and maybe some other sandboxing stuff which I don't have knowledge about), and Please copies files into the execroot.

The other difference is that Bazel's execroot is inside `/tmp` dir, and Please just has a `plz-out/tmp` directly inside the workspace.

**Comment by burdiyan** [comment]
Maybe a solution for this could be if Bazel could optionally copy files instead of symlinking them into the execroot?

**Comment by alexeagle** [comment]
Since the sandbox-exec is deprecated, https://github.com/bazelbuild/bazel/issues/8230#issuecomment-921074074
someone needs to make a new sandbox. The comparison with please is useful.
@hicksjoseph or @sventiffe is there a plan on the roadmap to deal with that deprecation?

**Comment by burdiyan** [comment]
I think there're indeed some lower-hanging fruits to improve Bazel on macOS situation.

First of all I'm going to assume (and I might be ridiculously wrong about it) that for a lot of people when they talk about sandboxed build, the mostly care about input files isolation rather than network and other stuff like that, which a true sandbox gives you. This is true for me as well: I don't care much about rules being able to access the network, because I know what the cost of it and I'm not going to do it, but I do want input files isolation, so that I know that I'm not missing to specify any inputs when I'm building the target.

If that is true for many people (at least on macOS), then Bazel could have several options to improve their lives (I know nothing about how complex implementing any of them could actually be):

1. Make `darwin-sandbox` strategy only care about input files. So stop using `sandbox-exec`, and simply assemble execroots with input files only for each target. Need to measure whether copy or symlink works better here, and maybe this even could be a flag to choose. Don't know if this could be really called a "sandbox" in this case though.
2. Change the understanding of what `local` strategy is. Right now `local` targets have access to all the workspace files, and it is probably not something that most people want. Maybe Bazel could do the same as Please for its local strategy, i.e. no other isolation except for input files. Again copy vs. symlink could be benchmarked, or even be configurable.
3. Do the same as previous, but making it a new type of strategy, e.g. `local-isolated` or whatever.
4. Indeed develop a proper low-level sandboxing facility for macOS, which is probably a non-option because of how complicated or even impossible this may be.

Basically options 1-3 are all the same, but named differently, and might have different level of impact in terms of implementation.

**Comment by meisterT** [comment]
> Indeed, I have replicated the same target in Please, and it takes 9 seconds in Please what takes 58 seconds in Bazel.

Is that with `experimental_reuse_sandbox_directories`?

**Comment by burdiyan** [comment]
> Is that with `experimental_reuse_sandbox_directories`?

@meisterT That's right. It actually didn't matter at all using this flag or not, the build time stays the same. Actually, preparing the sandbox directly looks like is very fast, it probably is just slow to read that many symlinks, or it's `sandbox-exec` which is causing it being slow.

**Comment by burdiyan** [comment]
So, I just made another test, which makes me think that it's indeed `sandbox-exec` that makes everything so slow.

With the same rule I mentioned above (many input files, building the whole Go app as a single target) I did the following:

1. I ran it with `--spawn_strategy=local`.
2. I passed all the input files as `SRCS` env variable to the build action.
3. Inside the build command I created `/tmp/bazel-sandbox` directory, and I copied all the input files in there.
4. Then I `cd /tmp/bazel-sandbox` and I built the app moving the result into the expected output place inside the execroot.

Basically I prepared my own little sandbox with only the input files specified inside the actual build rule.

Then it performed as expected and it took 6-7 seconds (the app uses CGO, that's why it's normal for it to be somewhat slow).

Then I did the same test but symlinks instead of copy. And it performs the same or even faster sometimes.

So, it's probably `sandbox-exec`, and not the symlinks being slow.

Again, this is all very quick and dirty tests, and I might be very wrong about all that, but still, I'd argue that Bazel should have a spawn strategy that would be just like "local", but would isolate the input files.

**Comment by meisterT** [comment]
Thanks for the details. @larsrc-google is the right person to look into this (after the holidays) and perhaps spawn a separate issue for `sandbox-exec`.

**Comment by larsrc-google** [comment]
This seems a non-trivial area, since Apple really wants everyone to build standalone apps using Xcode, and Bazel doesn't fit into that at all. I looked at the App Sandbox road, but it doesn't look feasible, all the tools need to be embedded (https://developer.apple.com/documentation/xcode/embedding-a-helper-tool-in-a-sandboxed-app).

jmmv@'s comment earlier indicates that sandbox-exec itself is not slow. And now you show that the symlinks are not slow. Maybe symlinks are only slow from within sandbox-exec? 

I don't have a target at hand that shows the extreme difference between local and sandboxed, so testing is difficult, but there is the CopyingSandboxedSpawn (currently used for Docker only) which could be used to test this hypothesis.

**Comment by jmmv** [comment]
I could never replicate any kind of slowdown with sandbox-exec or symlinks when microbenchmarking them in isolation. Based on that and various anecdotes, what I think is happening is that a *parallel* workload of file (symlink) creation and, especially, deletion bottlenecks in some file system lock. (If you look around, you'll find various mentions that HFS+ used to have some big locks in the kernel, and it's likely that APFS has them too.)

One thing I did *not* try was to compare a sandboxed build vs. a non-sandboxed one with --jobs=1 and check what the overhead looks like. That could maybe rule out this theory about lock contention...

**Comment by burdiyan** [comment]
I prepared a repo trying to replicate the scenario I was talking about previously. Hope it could be helpful: https://github.com/burdiyan/bazel-slow-sandbox-repro

**Comment by keith** [comment]
In this example since you're running golang outside of a bazel rule it's likely generating its own cache, which it is blocked from reading / writing to when using the sandbox.

**Comment by burdiyan** [comment]
🤦‍♂️ You're right! I had the feeling that it could be something like that, but I thought that I would get an explicit error of being unable to write or read the cache. But apparently Go compiler silently ignores the cache, or something like that. It is indeed most likely what you say @keith.

I still do think that Bazel should have a spawn strategy that would provide only file-level isolation and otherwise being exactly like local strategy. Symlinks or copies is probably not *that* important.

**Comment by larsrc-google** [comment]
I don't see any great advantage to such a strategy, certainly not strong to merit the maintenance and extra complexity. We have too many sandboxing strategies already.

**Comment by ob** [comment]
@burdiyan, I've fallen for [something similar before](https://github.com/bazelbuild/bazel/issues/7527#issuecomment-567264108) where `clang` was super-slow in sandboxed mode (20x slower) and it was because of the module cache being recreated all the time, not due to any particular slowness of `sandbox-exec`. Like @jmmv said, `sandbox-exec` is kind of an easy target, being unsupported and all, but actual benchmarks have never shown it to be particularly slow.

Just for kicks I tried building a small part of the LLVM compiler and benchmarked the following scenarios:

1. local strategy
2. darwin-sandbox strategy
3. darwin-fake-sandbox strategy that uses a no-op `sandbox-exec` that just calls `execvp(3)`
4. darwin-copy-sandbox strategy that uses the `CopyingSandboxedSpawn` mentioned by @larsc-google in a comment above.

The code is pushed to the `darwin-fake-sandbox` branch [here](https://github.com/ob/bazel/tree/darwin-fake-sandbox), I didn't bother pluggin it to the build so if you want to try it you need to manually compile the `fake-sandbox-exec` program an change the path in the code.

These are the results:

```
❯ hyperfine -p "~/o/bazel/bazel-bin/src/bazel --batch clean" "~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=local --config=generic_clang @llvm-project//clang:clang-tblgen" "~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen" "~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-fake-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen" "~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-copy-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen"
Benchmark 1: ~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=local --config=generic_clang @llvm-project//clang:clang-tblgen
  Time (mean ± σ):     38.560 s ±  1.027 s    [User: 380.953 s, System: 31.037 s]
  Range (min … max):   36.691 s … 39.787 s    10 runs

Benchmark 2: ~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen
  Time (mean ± σ):     41.092 s ±  0.826 s    [User: 398.602 s, System: 54.100 s]
  Range (min … max):   39.875 s … 42.639 s    10 runs

Benchmark 3: ~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-fake-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen
  Time (mean ± σ):     40.776 s ±  0.741 s    [User: 391.842 s, System: 53.463 s]
  Range (min … max):   39.502 s … 42.046 s    10 runs

Benchmark 4: ~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-copy-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen
  Time (mean ± σ):     53.806 s ±  1.171 s    [User: 445.051 s, System: 115.300 s]
  Range (min … max):   52.555 s … 56.255 s    10 runs

Summary
  '~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=local --config=generic_clang @llvm-project//clang:clang-tblgen' ran
    1.06 ± 0.03 times faster than '~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-fake-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen'
    1.07 ± 0.04 times faster than '~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen'
    1.40 ± 0.05 times faster than '~/o/bazel/bazel-bin/src/bazel --batch build --spawn_strategy=darwin-copy-sandbox --config=generic_clang @llvm-project//clang:clang-tblgen'
```

As you can see, removing the `sandbox-exec` command is a wash and copying is slower than symlinking (news at 11).

**Comment by larsrc-google** [comment]
Thank you for that nice investigation, @ob! It matches my expectations nicely. It's possible that copying very small files is faster, but at 6% speed difference it's not a high priority.

**Comment by jmmv** [comment]
Indeed, thanks for bringing that up and with proper numbers, @ob !

The shared compiler caches are a problem with the sandbox. I recall hacking the ObjC rules back when I was looking into this to support reusing the cache even when sandboxing is enabled, in an attempt to get more realistic measurements of the overhead. The problem was that the builds I was playing with didn't work properly with sandboxing enabled (IIRC) so I could never get any sane numbers...

I know reusing these caches goes against the spirit of sandboxing, but Bazel should probably allow this (by default) for performance reasons.

(And I very much doubt that traversing the symlinks per se is a problem. That information is likely going to be in the kernel's cache anyway. The individual syscalls to create and delete every symlink are much more costly, and using copies instead of symlinks has the exact same scalability problems here, even if they are COW.)

**Comment by burdiyan** [comment]
It turns out that Bazel already has a spawn strategy that I wanted. It only provides input isolation with symlinks, but doesn't offer any real sandboxing, AFAIK. It's called `processwrapper-sandbox`, and apparently it's semi-documented. CLI flags and reference docs don't mention anything about it, but I found it looking through the codebase, and it's also mentioned in the [sandboxing docs](https://bazel.build/docs/sandboxing).

I tried it, and apparently it does exactly what I want.

As far as using this with Go, it turns out that Go build tool uses file path as a cache-key somehow, that's why sandboxed builds never reuse Go build cache, because the path always changes as Bazel generates numerical increasing sandbox ID for the sandbox directory, and it's different for every run.

I solved it by copying the sandboxed dir into a different directory with a stable path, and now it reuses Go build cache perfectly. Hope someone finds it helpful.

I wonder if `processwrapper-sandbox` should have better docs?

Also, Is there a way to avoid sandboxed builds use different paths for each run? Maybe instead of a number it could be name of the target or something like that?

**Comment by larsrc-google** [comment]
If you want to have things cached between builds (apart from Bazel's cache), you should use a [persistent worker](https://bazel.build/docs/persistent-workers) instead of trying to make the sandbox less sandboxy. I'm not finding a worker implementation right off the bat, but then I'm not a Gopher, so I might not be looking in the right places. Workers can also use sandboxing, but will have a fixed path per worker.

**Comment by burdiyan** [comment]
@larsrc-google Thanks for the advice!

I've been thinking about persistent workers too, but didn't find anything existing. And for my rather simple use case I'm OK with trading off some hermeticity/reproducibility/correctness for simplicity and better interop with existing tools and language ecosystem.

I believe a lot of projects of non-Google scale, which currently use Make or custom scripts could benefit from such approach. If one is being careful with remote/disk caching, and doesn't need remote execution, Bazel could be a great alternative to messy Makefiles and ad-hoc scripts, IMHO.

**Comment by mistic** [comment]
@larsrc-google @philwo do we have any developments on this front? Is this being prioritised somehow? I'm using Bazel under a `Node.js` with `rules_nodejs` on a project that depends on a considerable amount of `node_modules` and the `macOS` sandbox is really hurting us. As we keep adding code and dependencies it keeps getting worse. We don't want to use the `local` strategy as the isolated dependencies build is one of the reasons we move into Bazel in the first place. I really believe in the absence of a `macOS` sandbox with decent performance we should consider to introduce a new strategy, as suggested above, `local-isolated` (or maybe `pseudo-sandbox`) as it will cover one of the biggest benefits of sandboxing while allow to have a decent performances on `macOS`. Having that stop gap in place will allow for more room to investigate around a decent sandbox for macOS or even develop one. On top of that we also don't have any guarantee `Apple` won't just remove `sandbox-exec` on the next macOS upgrade (as it is already deprecated, just look what happens recently to `python2` in a minor release). When summing up all of this I think we should consider this as a high priority thing on Bazel.

**Comment by keith** [comment]
Note that as far as `sandbox-exec` being deprecated goes, Apple is using it in the Swift package manager here https://github.com/apple/swift-package-manager/blob/e25a590dc455baa430f2ec97eacc30257c172be2/Sources/Basics/Sandbox.swift#L37 so I assume if it goes away they will also need a replacement,.

**Comment by mistic** [comment]
@keith Thanks for the inputs! I'm not very used to iOS or macOS development overall but what I have been reading about the theme it looks like the initially purposes by which apple created `sandbox-exec` are mostly covered by the new `App Sandbox` now. I really think Apple could just decide to remove it in one of the next upgrades as all of their internal needs are covered other way. Overall I think the outlined arguments above are still valid and important: we are heavily relying on a piece of software which is now deprecated for 2 years, without any further guarantees of support, and even worst than the previous which is just giving us all a really poor build performances. I believe that should be prioritised to improve the overall Bazel experience for all using it under `macOS`.

**Comment by jmmv** [comment]
@mistic As has been mentioned a few times already, sandbox-exec *is not* why macOS sandboxing is slow. The reason sandboxing is slow is for two reasons: first, the creation *and* deletion of symlinks to create each sandbox; and second, the fact that some actions create local caches that cannot be reused across actions. The latter is by far the largest performance penalty. I haven't checked, but I suppose the nodejs rules do this.

Also note that, even if Apple claims that sandbox-exec is deprecated, even macOS still uses it. Last time I checked (late 2019, see https://jmmv.dev/2019/11/macos-sandbox-exec.html), it was used for many system internal services. I do not have access to a macOS system right now to check how things are these days.

**Comment by mistic** [comment]
Hi @jmmv thanks for sharing this with me!

I have been profiling and collecting a lot of stats on these lately and I don't think what I'm seeing is completed validated by what you've mentioned.

> and second, the fact that some actions create local caches that cannot be reused across actions. The latter is by far the largest performance penalty. I haven't checked, but I suppose the nodejs rules do this.

I have been running the same exact codebase, same build in the same machine both on `macOS` and `ubuntu`. On rules nodejs v5 more actions are being created around the `node_modules` inputs than in the previous version and we definitely see a huge penalty cost on `macOS`. However that does not replicates on unix as expected according what you were mentioning. On `macOS` the performance penalty is it is being `~3.5x` times slower than before increasing from `280s` to `950s`. On the other side, same test on Ubuntu, has only showed a performance penalty of `~1.3x`. Executing with `local` strategy get everything back to normal but is something we don't want as we will loose the benefit of having isolated `src` dependencies during the build and even the detection of missing dependencies during development. 

> first, the creation and deletion of symlinks to create each sandbox; 

For that one I don't have doubts it could be it and thats why I think there is something needed to be done around this. If we can have a `local` like strategy that simulates the `srcs` dependency isolation during the builds that would be a really good stop gap while a better solution isn't found.

**Comment by jmmv** [comment]
@mistic I don't know enough about nodejs behavior so I cannot comment much on that. All I can say is that the massive slowdowns that have been previously observed in the Go and ObjC rules were because of 1. local caches being lost across actions and 2. some actions mapping huge SDKs into each action's execution root (the Go rules do that and thus you end up creating thousands of links for every action). For comparison, if you try to build Bazel itself with and without sandboxing on macOS, you'll notice that the performance differences are negligible -- so sandboxing itself on macOS is quite efficient if the rules cooperate.

Regarding symlinks, I'd suggest looking at https://blog.bazel.build/2017/08/25/introducing-sandboxfs.html and other stuff I did around sandboxfs. I never got to productionize that because I hit some dead ends (FUSE going away on macOS and a very gnarly problem I described in https://jmmv.dev/2020/01/osxfuse-hardlinks-dladdr.html).

But I haven't been in the team for almost two years now... and haven't been using Bazel either, so my info is probably stale. I just have to chime in every time someone blames sandbox-exec because I just don't believe that the problem is there 🙃

**Comment by larsrc-google** [comment]
@mistic Just to check: You are running with `--experimental_reuse_sandbox_directories`, yes?

**Comment by mistic** [comment]
@larsrc-google yes I'm using it as without that flag the performance decrease is even worse

**Comment by larsrc-google** [comment]
Ouch. And the Linux/Mac difference you mention may indeed indicate that the cache is not the main problem. Like Joe, I don't know that much about nodejs behaviour, but would be willing to investigate if you can point at some code that I can reproduce it with. You should also be able to see slow setup time - if that's the problem - by using [the profiler](https://bazel.build/rules/performance#performance-profiling) - don't forget to pass `--noslim_profile`.

**Comment by mistic** [comment]
@larsrc-google Thank you for your availability to help investigating this. In order for your to test around with what I have been able to identify you can use the following branches:

**Branches:**
- rules v4.7.0: https://github.com/mistic/kibana/tree/main
- rules v5.4.0: https://github.com/mistic/kibana/tree/hold-upgrade-rules-v5

As states before the difference between those branches is the `rules-nodejs` version in use. The `v4` are really faster than `v5` and the difference I was able to found so far is that they use a really smaller amount of actions when dealing with `node_modules` dependencies. So far I was also been able to track the performance decrease to the following commit https://github.com/bazelbuild/rules_nodejs/commit/91fa0ae68ce2c2d1fd59323ca1bed5fbbb64cf99#diff-15228457773ae15b53a9887c143a90a63eec7c19950ba30a37b541e3275d82f7R1055

**Steps to reproduce:**
1 - checkout each of the branches
2 - make sure you have yarn@1.21.1 and node@16.14.2 installed and globally available on your path. (I just using nvm to manage node and installing yarn with `npm install -g yarn@1.21.1`)
3 - then just run `yarn kbn bootstrap` . That command is just almost equivalent as just running  `yarn bazel build //packages:build` in functionality but with some useful setup for the project so I recommend you to run the former instead.
When switching between branches please run `yarn kbn reset` as it will delete node modules and run `bazel clean --expunge`. You will be able to experience what I’ve reporting in terms of performance decreasing when comparing the times it takes to `yarn kbn bootstrap` or `yarn bazel build //packages:build` between `v4` and `v5` having in account you have run `yarn kbn reset` before those to assure nothing is cached in place.

**Comment by tony-scio** [comment]
I'm curious if there's a canonical set of recommended flags for nodejs_rules+macos.

The sandbox is hurting us pretty severely. Here's an example of one test that takes 31s outside of bazel, but 72s inside it (it's typical of our tests). More than 2x overhead!

<img width="930" alt="Screen Shot 2022-05-17 at 12 48 34 PM" src="https://user-images.githubusercontent.com/49661492/168898502-e89f9b7f-332a-4b27-96bd-f19fd7ece150.png">

Breakdown of overhead:
- 3s `fsvc.getDirtyKeys`
- 3s `getDirtyActionValues.stat_files`
- 8s `AbstractSpawnStrategy.getInputMapping`
- ~10s Untraced
- 17s `sandbox.createFileSystem`
- 6s `sandbox.delete`

These flags didn't help significantly:
```
build --spawn_strategy=processwrapper-sandbox
build --experimental_reuse_sandbox_directories
build --experimental_sandbox_async_tree_delete_idle_threads=1
```

Nor did `exports_directories_only = True,` in our `yarn_install()`.

`--spawn_strategy=local` largely fixes the issue for nodejs_rules, but breaks some targets we have in other languages and I'm skeptical about using it for reasons mentioned elsewhere in the thread.

Our rules are:
```python
# jest_test is a copy of jest.bzl from the nodejs_rules repo
# test_dependencies is list of about 30 `@npm//...` targets

filegroup(
    name = "core",
    srcs = glob(
        [
            "core/**/*.ts*",
        ],
        exclude = [
            "**/*.spec.ts*",
        ],
    ),
)

jest_test(
    name = "core_tests",
    size = "large",
    srcs = glob([
        "core/**/*.spec.ts*",
    ]),
    jest_config = "jest.config.js",
    deps = test_dependencies + [
        ":core",
    ],
)
```

Hoping for help from someone and happy to provide help in return in terms of repros, experiments, etc.

**Comment by tony-scio** [comment]
In Bazel 6.2 we continue to have to keep the sandbox disabled for tests on osx due to performance overhead of 2x (particularly for python and typescript which have large numbers of runfiles).

**Comment by larsrc-google** [comment]
It might be worth parallelizing some of these things, especially once Loom makes it cheaper to use threads. Having a small threadpool for creating symlinks ought to help. But fundamentally, OS X is just much  slower than Linux at handling lots of creation and deletion of small files.

**Comment by meisterT** [comment]
@tony-scio is your overhead of 2x with the flag `--experimental_reuse_sandbox_directories`?

**Comment by oquenchil** [comment]
> These flags didn't help significantly:
> build --spawn_strategy=processwrapper-sandbox
> build --experimental_reuse_sandbox_directories
> build --experimental_sandbox_async_tree_delete_idle_threads=1

@tony-scio in most cases `--experimental_reuse_sandbox_directories` should help significantly. Would it be possible for you to share a small example workspace where sandboxing enabled gives you a 2x overhead? I see your `core` and `core_tests` target but just with that it will be hard to reproduce.

**Comment by DavidZbarsky-at** [comment]
@oquenchil I ran into this issue as well, posting in here to continue the discussion, but let me know if you'd prefer a separate issue.

We are using `rules_js`. We depend on quite a few npm packages, which the JS rules expose as TreeArtifacts (so I believe they are only 1 symlinked per package). We have unit tests that takes 1-2 seconds to run outside Bazel, but significantly longer under Bazel. Here's a stripped-down repo with a single `sh_test` depending on node modules: https://github.com/DavidZbarsky-at/nodejs-repro/blob/stuff/BUILD.bazel. I believe there are around 15K entries in the runfiles manifest for this test action.

On my (M1 mac) machine, the sh_test spends around 6-7 seconds in RepoMappingManifestAction (@fmeum fixed this already), and 6 seconds in sandbox.createFileSystem. Running the same test in a docker image on the same machine shows only 1 second in sandbox.createFileSystem.

I have reuse_sandbox_directories enabled, but running the test several times in a row shows the same timing each time.

Even stranger; I tried with `--noenable_runfiles` flag as well, and confirmed that the runfiles were not populated, but I still see the same performance numbers. I would have expected it to be faster if we don't need to create symlinks. Perhaps I'm confused how this is supposed to work?

**Comment by oquenchil** [comment]
Hi David,

Just to make sure the slowness is coming from sandboxing, if you run that same target with `--spawn_strategy=local` and `---spawn_strategy=sandboxed`, do you see the massive difference?

Does your repro build with bazel at head? If not, what's the latest version that it works with?

Thanks for the repro, I will take a look in the following days!

**Comment by larsrc-google** [comment]
The Linux file system is _much_ more optimized for the kinds of operations that happen in builds.

**Comment by coeuvre** [comment]
@DavidZbarsky-at Hi David, I see there are `context.prefetchInputs` spans in the first screenshot which take a couple of seconds. I am wondering did you use any remote or disk cache in the examples (maybe accidentally with `~/.bazelrc`)?

**Comment by DavidZbarsky-at** [comment]
@coeuvre yes there was a remote cache (though this multisecond gap is maybe another performance issue?). I see similar timing without it on the sandboxed tests, though it appears to fix the unattributed time in the `local` strategy:

`darwin-sandbox`
<img width="1083" alt="image" src="https://github.com/bazelbuild/bazel/assets/94403429/cf6e3b28-8757-46ea-bd77-bc5d1ac70809">

`local`
<img width="996" alt="image" src="https://github.com/bazelbuild/bazel/assets/94403429/1f533dc7-068d-4d09-afa4-cf3190a630a2">

**Comment by coeuvre** [comment]
@DavidZbarsky-at Thanks for confirming. Yes, it's another performance issue. Created https://github.com/bazelbuild/bazel/issues/20555 to track.

**Comment by oquenchil** [comment]
There is another performance issue: https://github.com/bazelbuild/bazel/issues/20584

I will take care of this one. After it's checked in I will profile again and see what else we can do.

**Comment by oquenchil** [comment]
@DavidZbarsky-at there are several things we can do to improve performance. I wanted to ask you a couple of questions though:
1. Would you mind if the overhead is gone only in incremental reruns of the test but the first time you run it the `createFileSystem` overhead is still there?
2. You have thousands of files as inputs. Are they spread out over dozens of directories or are they concentrated in individual directories? If you can provide some real world statistics for your JS builds that would be helpful, something like 10000 files in total in 700 directories, not a single directory with more than 1000 files, etc...

I will be able to take a look again on the 2nd of January but any additional information you can provide would be very helpful.

Thanks!

**Comment by DavidZbarsky-at** [comment]
@oquenchil Awesome, that sandbox reuse one seems like it will be a big help for us. Thanks!

Re: your questions:

1. That's better than the current situation but not ideal. In particular I'd like to be able to run a set of 20-30 tests locally before going to CI, and that becomes tricky with this overhead (i.e. 1-2s per tests * 30 / 4 cores is 15-20 seconds, but the sandboxing overhead here is going to be 3x+)
2. Looks like a power law to me. For our repo, I counted the number of files in each NPM package in node_modules (raw data attached in case you are curious). **Not every action has every NPM package as an input, but hopefully this gives you a sense of the kinds of things the NPM ecosystem does** Summary stats:

>>> sum(counts) # Total number of files
192197
>>> len(counts) # Number of directories
4197
>>> sum(counts)/len(counts)
45.7939004050512

>>> statistics.quantiles(counts, n=20)
[4.0, 4.0, 4.0, 5.0, 5.0, 5.0, 6.0, 6.0, 7.0, 8.0, 9.0, 11.0, 12.0, 15.0, 18.0, 24.0, 34.0, 56.0, 137.0]
(p5....................................................................................................................................p95)

[npm_packages.txt](https://github.com/bazelbuild/bazel/files/13721999/npm_packages.txt)

Happy to poke around and provide whatever else would be helpful, just let me know!

**Comment by larsrc-google** [comment]
Reducing unnecessary dependencies is always good. But here's a thing I've been wanting to try but not been able to: With Loom virtual threads, could we do more parallelism in creating/destroying the sandbox?

**Comment by meisterT** [comment]
We are unfortunately not there yet with using JDK21 features (such as Loom) but @oquenchil looked at speeding up sandbox creation by a good plain old executor service yesterday (and that triggered his question above).

**Comment by DavidZbarsky-at** [comment]
I came across [this doc/idea](https://docs.google.com/document/d/1bvvx6gx42YSyhhqxlxQvi6eZzF-0J9k3bmV8_NCapuo/edit#heading=h.i71k1q1nlyuu), not sure if it's feasible but it seems like it would solve this case (folders with hundreds + or thousands of files) quite nicely

**Comment by larsrc-google** [comment]
I had been considering `io_uring`, but apparently it's not very secure yet. But I recently learned that `kqueue`, the Mac (BSD) equivalent is over a decade old and thus probably in a better state. Maybe that's why the default IO on Mac is relatively slow, serious IO users use `kqueue`? Anyway, a good proof-of-concept for that would be implementing `deleteTreesBelow` using `kqueue`, if that is promising we could add other APIs amenable to `kqueue` - and preferably also to Loom and maybe eventually `io_uring` - such as `createDirsInDir` and `createSymlinksInDir` (strawman names).

**Comment by oquenchil** [comment]
> I came across [this doc/idea](https://docs.google.com/document/d/1bvvx6gx42YSyhhqxlxQvi6eZzF-0J9k3bmV8_NCapuo/edit#heading=h.i71k1q1nlyuu), not sure if it's feasible but it seems like it would solve this case (folders with hundreds + or thousands of files) quite nicely

The idea as I originally described it on that document doesn't quite work because when you symlink directories, the path `foo/bar/..`  doesn't take you to `foo` but to the parent of the directory pointed to by `foo/bar`. After considering several ideas bind mounting seems like it has potential and we are still looking into it but that's only available on Linux.

That leaves us with macOS where we are at the point of simply optimizing the existing strategies instead of coming up with a new one entirely. As meisterT pointed out, me adding multi-threading is what prompted those questions.

What I found was that with multi-threading the bottleneck is in the directory lock held by the kernel when writing a file. If most files are concentrated in a small handful of directories then multi-threading won't help much but looking at the raw data you provided it seems that's not the case, so it should cut down the time a bit.

Because of where I think the bottleneck is I don't think it will matter whether we use Loom.

 For cleaning up an existing stash, I tried moving a tree out of the way then deleting asynchronously. This also gives a significant speed bump. I will check in those two optimizations then we can see where we are at. Without symlinking directories or bind mounting, we might hit a limit in how much more we can do on macOS for cold builds but we aren't there yet.

**Comment by rickeylev** [comment]
Python builds also suffer from the performance hit due to so many files needing to be materialized. The basic issue is every Python program needs a runtime, and a Python runtime installation is a lot of files (2000 to 4000 files). This IO costs adds up. I think it was Tensorflow that contacted me asking for advice because their CI builds went from 45 minutes using the system python to 2+ hours using the downloaded runtimes, and all evidence pointed to all the extra IO work creating thousands of symlinks for every test. There's similar reports to rules_python by other users, too, so its not just them.

While we think we can partially avoid this by zipping up most of a runtime installation (we can't zip up all of it), that feels like a lateral move overall. There is a still an O(2000 - 4000) cost paid to create the zip file and the cost of copying a decent sized file around. That's still a lot more cost than creating a single symlink to a directory. Zipping stuff correctly can also be somewhat subtle (e.g., making sure you're avoiding nondeterminism bugs).

Relatedly, for Bazel CI and Python, it currently gets to avoid this issue because it's using the system runtime. However, my 2nd or 3rd todo list item right now is to upgrade Bazel CI's rules_python version, which will make it use the downloaded runtimes; i.e. every python build action or python test it runs will incur a separate 2,000+ file copy.

Relatedly, pypi Python dependencies also suffer from this problem, as they work mostly the same as the runtime. A repo rule downloads a zip file and extracts it. The program needs those files at runtime, so they all get copied over into runfiles. But all it really needs is a symlink'd directory pointing to the files. The footprint of third party dependencies is easily the size of the runtime itself or orders of magnitude larger. As an example, I have a program with 5 direct dependencies, and it turns into about 15,0000 runtime files (technically ~7k, but they get duped because of the legacy-externals-runfiles behavior). In comparison, if directories were symlinked, it would be about 32 symlinks. Unfortunately, trying to zip those up isn't as easy, straight forward, or feasible.

**Comment by larsrc-google** [comment]
> What I found was that with multi-threading the bottleneck is in the directory lock held by the kernel when writing a file. If most files are concentrated in a small handful of directories then multi-threading won't help much but looking at the raw data you provided it seems that's not the case, so it should cut down the time a bit.

That actually suggests an API for building an entire tree of symlinks. Then for MacOS, the implementation could spread the writing across multiple directories to get around the directory lock bottleneck.

**Comment by oquenchil** [comment]
>That actually suggests an API for building an entire tree of symlinks. Then for MacOS, the implementation could spread the writing across multiple directories to get around the directory lock bottleneck.

I talked about symlinking directories a while ago [here](https://docs.google.com/document/d/1bvvx6gx42YSyhhqxlxQvi6eZzF-0J9k3bmV8_NCapuo/edit#heading=h.ecbf1pz2pxdd). I sent that to [bazel-discuss](https://groups.google.com/g/bazel-discuss/c/K2M93UE1cQY/m/WA5txoVsJgAJ).

The reason why symlinking directories doesn't work is because as soon as you have `symlink_dir1/symlink_dir2/../symlink_di3/`. The `..` won't behave like you expect, it will go to the parent directory of the target of the link.

There are alternatives using bind mounts on Linux which we are currently looking at but that won't solve the problem for macOS which is what this issue is about.

Regarding the directory lock bottleneck, that wouldn't solve it since each directory would still contains as many files as before. In any case it doesn't look like in practice it's as much of a bottleneck at least for JS since files are not grouped in a single directory. Not sure in the case of Python.

> But all it really needs is a symlink'd directory pointing to the files.

For the reason I mention above we wouldn't want to indiscriminately use the strategy of symlinking directories. However,  if the rule author could guarantee that there would never be a `..` after the symlinked directory then this could be special cased. The sandboxing code in Bazel couldn't take these shortcuts and make these decisions on its own but we could add a mechanism for the cases where the rule author knows what they are doing. If they aren't careful, they might expose themselves to the maintenance burden of user builds breaking in not very clear ways when they do `..` from within an input file passed to the action. I saw this type of breakage in at least two different languages rulesets when I tried the symlinking directories idea.

Richard, please tell me how you envision such an API as passed from the rule or toolchain implementation and I can write a prototype.

**Comment by meisterT** [comment]
@oquenchil wrote this related doc: https://docs.google.com/document/d/1En33vAYIJlYxn1YovJVMEUHD48fqBLIm7-3ebpnjTkc/edit?tab=t.0#heading=h.6flzffxtmybg

and @TheGrizzlyDev had some interesting ideas how to use LD_LIBRARY_PRELOAD

**Comment by TheGrizzlyDev** [comment]
> @oquenchil wrote this related doc: https://docs.google.com/document/d/1En33vAYIJlYxn1YovJVMEUHD48fqBLIm7-3ebpnjTkc/edit?tab=t.0#heading=h.6flzffxtmybg
> 
> and @TheGrizzlyDev had some interesting ideas how to use LD_LIBRARY_PRELOAD

yeah, I'm currently working on it but it isn't ready to come out yet and it's mostly a bunch of experiments glued up together. MacOSX has a mechanism analogous to LD_PRELOAD that allows inserting DYLIBs prior to the execution of a binary. Then one can use mach-o's `__DATA,__interpose` section to inject replacements for functions like open/read/write/etc. which would allow virtualising the file system. 

My goal is a bit broader than that and I'd like to create more complete sandboxing primitives with the end goal of having macos containers that more closely resemble the linux ones, but it can also be used to mimic a behaviour similar to what sandboxfs used to do and with relatively low overhead. 

Such an approach wouldn't work on linux where it's rather common to statically link binaries, but with macosx you need to dynamically link at least against libSystem (more on that [here](https://forums.developer.apple.com/forums/thread/706419) ), which is where the symbols we care about reside.

I expect the first POC to come out in a few weeks, maybe around Christmas, when I will have more time to work on it. That won't be enough to solve this issue but hopefully it should be enough to prove the validity of this approach. I will keep working with @meisterT to figure out the next step. 

Hopefully this approach works, since I've bet a bazel-themed rubber duck on it :duck:

**Comment by TheGrizzlyDev** [comment]
I also wanted to point out how receptive @meisterT has been to my proposal. It really shows that, despite all the issue with MacOSX, the team is really trying their best to make it work and I really appreciate that.

**Comment by NathCn** [comment]
Hi! 

Is there any update on the topic? This performance issue is very noticeable on our team as well, but we're not sure the cause is this or some third-party interfering software (or both!).

**Comment by oquenchil** [comment]
Are you passing [`experimental_inmemory_sandbox_stashes`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/sandbox/SandboxOptions.java;l=342)?

Could you produce a [JSON trace profile](https://bazel.build/advanced/performance/json-trace-profile) to make sure the issue is sandbox performance?

**Comment by NathCn** [comment]
I didn't see this parameter before D: I just tried again with the parameter but it ended up taking the same time. I don't think I can share this profile since it contains paths/company IP-related info, but glancing at it it doesn't look like sandbox is actually the issue; not 100% sure on how to check that though.

I'll check with colleagues as well and generate a profile without the flag and compare, maybe they'll have more insight. It might be that third-party app...

**Comment by thesayyn** [comment]
> MacOSX has a mechanism analogous to LD_PRELOAD that allows inserting DYLIBs prior to the execution of a binary

I independently came to the same conclusion as you and then saw you comment here, i have a working PoC for this and it scales really well, but the issue is [SIP](https://developer.apple.com/documentation/security/disabling-and-enabling-system-integrity-protection), which disallows DYLD_* on apple signed binaries. 

So this strategy makes apple rules such as rules_apple, rules_swift unusable with this strategy.

**Comment by TheGrizzlyDev** [comment]
> > MacOSX has a mechanism analogous to LD_PRELOAD that allows inserting DYLIBs prior to the execution of a binary
> 
> I independently came to the same conclusion as you and then saw you comment here, i have a working PoC for this and it scales really well, but the issue is [SIP](https://developer.apple.com/documentation/security/disabling-and-enabling-system-integrity-protection), which disallows DYLD_* on apple signed binaries.
> 
> So this strategy makes apple rules such as rules_apple, rules_swift unusable with this strategy.

That's what I've been working on for quite a while now ahahahha. I have taken a temporary break but I'll probably resume this work soon-ish. There are a few ways to get around this issue:
- rebuild the tools from source, a bunch of these tools are opensource, so we can just rebuild them and drop them in a ficticious sysroot
- create wrapper for tools, these wrappers would translate arguments and set the env of a SIP protected tool before invoking it so that it acts the same way you'd expect it inside the sandbox
- use OSS alternatives of binutils

This should be a somewhat reasonable approach because the number of binaries protected by SIP we care about is somewhat limited. I will probably end up combining all of them. Finally, there is an uglier approach that works and might be a reasonable short term solution. If you copy a binary from a SIP protected location to somewhere that isn't SIP protected, then you unsign it and run it in x86_64 (either by extracting it with lipo or by virtue of emulation) then you can use DYLD_INSERT_LIBRARIES on it (arm64e adds some protections that make unsigning insufficient and SIP binaries just won't run after being moved). This works because macos is currently shipping universal binaries. So, why temporary? I fear that in the long run Apple will get completely rid of x86_64 support, making this policy of shipping universal binaries a thing of the past.

Anyhow, with this we should be able to use this apprach with rulesets that require native apple tooling. If you'd like to help I can whip the codebase into shape and drop some documentation, I'd love to have some contributors landing me a hand as I am doing this in my free time.

**Comment by thesayyn** [comment]
I came to the same conclusion as you did, however, i believe building all the SIP protected tools from source is a massive effort, and not all possible since things like xcodebuild might not be open. 

> If you'd like to help I can whip the codebase into shape and drop some documentation, I'd love to have some contributors landing me a hand as I am doing this in my free time.

I'd love to, i am on the Bazel slack, we can chat there if you are.

**Comment by Ashutosh0x** [comment]
@fenghaolw Hey! I've been looking into the sandbox slowness on macOS as well. I just opened #28673 to add profile caching, which should cut down on a lot of the per-action overhead we're seeing. Specifically helps when you have many small actions!
